apple/swift-docc
# Add near-miss suggestions for unresolved symbol link error messages

Bug/issue #, if applicable: This contributes to rdar://59660520 but doesn't fully resolve it.

## Summary

This adds a mechanism for sorting and filtering strings based on how "similar" they are to a specific string for use to create near-miss suggestions in diagnostics.

The changes are divided in two parts:
- One that wraps the [the difference capabilities in Swift][1] to calculate the length of the segments of common and difference elements. 
- Another that "score" these segments to "rank" values based on how "similar" they are to the authored value.

For the second part there is no single right or wrong answer for how to do that. The approach I took is to write test of what results I would expect when matching a collection of possibilities against a certain value and then tweak the "score" and "ranking" implementations until they produces the results that I'd expect. This implementation is completely arbitrary.

My hopes is that when there are situations where this doesn't behave well, we'll add those as more tests and tweak the "score" and "ranking" implementations again.

[1]: https://developer.apple.com/documentation/swift/array/difference(from:by:)

## Dependencies

None

## Testing

Please give this a spin with some real code bases to see how it performs with real links in real projects.

To do so:

1. Checkout this branch and build the `docc` executable.
2. Build documentation for another project with `DocCUseHierarchyBasedLinkResolver` enabled and `DOCC_EXEC` pointed to the locally built `docc` executable.

## Checklist

Make sure you check off the following items. If they cannot be completed, provide a reason.
